specs/agents/metrics-otel.md

```diff
@@ -0,0 +1,166 @@
+                double bounds = otelBoundaries.get(0);
+                if(bounds > 0) {
+                    bounds /= 2;
+                }
+                intakeValues.add(bounds);
+            } else if (i == bucketCount - 1) { //last bucket
+                intakeValues.add(otelBoundaries.get(bucketCount - 2));
+            } else { //in between
+                double lower = otelBoundaries.get(i - 1);
+                double upper = otelBoundaries.get(i);
+                intakeValues.add(lower + (upper - lower) / 2);
+            }
+        }
+    }
+    
+    return new IntakeV2Histo(intakeCounts, intakeValues);
+}
+```
+
+The same algorithm is used by the APM server to convert OTLP histograms.
+The `sum`, `count`, `min` and `max` within the OpenTelemetry histogram data are discarded for now until we support them in IntakeV2 histograms.
+
+## Labels
+
+Conceptually, elastic metric labels keys correspond to [OpenTelemetry Attributes](https://opentelemetry.io/docs/reference/specification/common/#attribute).
+When exporting, the agents MUST convert the attributes to labels as described in this section and MUST group metrics with the same attributes and OpenTelemetry instrumentation scope together into a single metricset.
+
+Attribute keys MUST be sanitized before being used as label keys: Every occurrence of the characters `.`, `*` and `"` must be replaced with a single `_`.
+
+The attribute values MUST NOT be modified and their type MUST be preserved. E.g. `strings` must remain `string`-labels, `booleans` must remain `boolean`-labels.
+Metricsets currently do not support array-valued labels, whereas OpenTelemetry attribute values can be arrays. For this reason, array-valued attributes MUST be ignored. Agents SHOULD emit a warning with a message containing the metric name when attributes are dropped because they are array-valued.
+
+To summarize and to give an example, given the following set of attributes
+
+| **Attribute Key** | **Attribute Value** | **Type**       |
+|-------------------|---------------------|----------------|
+|  `foo.bar`        | `baz`               | `string`       |
+|  `my_array`       | `[first, second]`   | `string array` |
+|  `testnum*`       | `42.42`             | `number`       |
+
+the following labels must be used for the resulting metricset:
+
+| **Label Key** | **Label Value** | **Type**       |
+|---------------|-----------------|----------------|
+|  `foo_bar`    | `baz`           | `string`       |
+|  `testnum_`   | `42.42`         | `number`       |
+
+The OpenTelemetry specification allows the definition of metrics with the same name as long as they reside within a different instrumentation scope ([spec link](https://opentelemetry.io/docs/reference/specification/metrics/api/#get-a-meter)). In order to avoid conflicts, agents MUST add a mandatory label to the metricset with the key `otel_instrumentation_scope_name` and the corresponding instrumentation scope name as label value.
```

Comment:
This is another difference to how the APM server currently does things when receiving metric data via OTLP.
The APM-Server will create a different metricset per `instrumentation scope`, but won't add any label to distinguish them (to my knowledge).
This means that when displaying the metrics in Kibana, the metrics with the same name will get mixed up even if they come from different instrumentation scopes.

Would be great if @elastic/apm-server could clarify here if I'm missing something or if my proposed label makes sense and we maybe should add it on the OTLP intake side aswell.

Reply:
Some notes on this.

1. https://opentelemetry.io/docs/reference/specification/metrics/api/#get-a-meter and https://opentelemetry.io/docs/reference/specification/glossary/#instrumentation-scope are not very clear, but one *could* read from those that the instrumentation scope `version` (and maybe `schema_url`) fields may also be relevant for uniquely identifying a scope.
2. FWIW, looking at the Prometheus exporter from the OTel JS SDK, the instrumentation scope is not included anywhere in the output. I created a JS script that has two `my_counter` metrics with diff instrumentation scopes:

```js
...
const meter = otel.metrics.getMeter('my-meter') // 'my-meter' is the instrumentation scope name
const counter = meter.createCounter('my_counter', { description: 'My Counter' })

const meter2 = otel.metrics.getMeter('my-meter2')
const counter2 = meter2.createCounter('my_counter', { description: 'My Counter 2' })

setInterval(() => {
  counter.add(1)
  counter2.add(1)
}, 1000)
```

The resulting exported Prom metrics:

```
% curl localhost:3002/metrics
# HELP target_info Target metadata
# TYPE target_info gauge
target_info{service_name="use-otel-prom",telemetry_sdk_language="nodejs",telemetry_sdk_name="opentelemetry",telemetry_sdk_version="1.9.1"} 1
# HELP my_counter_total My Counter
# TYPE my_counter_total counter
my_counter_total 317 1679701732718
# HELP my_counter_total My Counter 2
# TYPE my_counter_total counter
my_counter_total 317 1679701732718
```

So, I wonder if this is the conflict we could later come back if/when there is a real use case.

3. If we were to add an intake field or label for this, https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/mapping-to-non-otlp.md#instrumentationscope describes `otel.scope.name` and `otel.scope.version`. I'm not sure if that applies here.

Reply:
> https://opentelemetry.io/docs/reference/specification/metrics/api/#get-a-meter and https://opentelemetry.io/docs/reference/specification/glossary/#instrumentation-scope are not very clear, but one could read from those that the instrumentation scope version (and maybe schema_url) fields may also be relevant for uniquely identifying a scope.

It is true that the instrumentation scope name alone might not be good enough to prevent clashes in all cases. It is good enough to prevent cross-library clashes (e.g. library A and B happen to export a metric with the same name) but won't handle the case where you have multiple versions of the same library with breaking changes in their metrics.

My thought here was to start simple and to leave the option open to maybe just use a single label (name and version concatenated) instead of adding both immediately. 

> FWIW, looking at the Prometheus exporter from the OTel JS SDK, the instrumentation scope is not included anywhere in the output. I created a JS script that has two my_counter metrics with diff instrumentation scopes
> So, I wonder if this is the conflict we could later come back if/when there is a real use case.

Interesting observation that this is currently not even handled by the Otel prometheus exporter.
We can definitely just ignore the `InstrumentationScope` for now and add a label later on, as it would be a non-breaking change.

We should however imo keep in the spec that InstrumentationScopes have to use a different metricsets, even if they have the same labels. Otherwise we would need to decide on a "winner" metric if we encounter a clash.

Reply:
In the meeting today we discussed using `otel.scope.name` and `otel.scope.version` and perhaps adding those at top-level fields (@gregkalapos volunteered).

@Mpdreamz mentioned (if I understood correctly) that metricbeat uses ECS `observer.name` (https://www.elastic.co/guide/en/ecs/current/ecs-observer.html#field-observer-name) for similar purpose?

Reply:
I've left the concrete addition of the fields for `otel.scope.name` and `otel.scope.version` for the future, so that this doesn't block merging this spec:
> Agents MUST report metrics from different instrumentation scopes in separate metricsets to avoid naming conflicts at collection time. This separation MUST be done based on the instrumentation scope name and version. In the future, we might add dedicated intake fields to metricsets for differentiation them based on the instrumentation scope identifiers.